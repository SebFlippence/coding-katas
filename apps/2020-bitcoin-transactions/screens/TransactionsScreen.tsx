import React, { useState } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { ScrollView, FlatList } from 'react-native-gesture-handler';

import TransactionRow from '../components/TransactionRow';
import { Transaction } from '../types/Transaction';
import { api } from '../constants/config';
import { useNavigation } from '@react-navigation/native';

export default function TransactionsScreen() {
  const [transactions, setTransactions] = useState<Transaction[]>([]);
  const { addListener } = useNavigation();

  addListener('focus', () => {
    async function fetchData() {
      const response = await fetch(`${api.hostname}${api.transactions}`);
      if (!response.ok) {
        throw new Error('There was a problem loading your transactions');
      }
      setTransactions((await response.json()).data);
    }
    fetchData();
  });

  const balance = transactions.reduce(
    (prev, curr) => ({
      amount: `${(parseFloat(prev.amount) + parseFloat(curr.amount)).toFixed(2)}`,
      currency: curr.currency,
    }),
    { amount: '0', currency: '' }
  );

  return transactions.length ? (
    <ScrollView style={styles.container} contentContainerStyle={styles.contentContainer}>
      <View style={styles.balanceContainer}>
        <Text style={styles.balanceText}>
          Balance: {balance.amount} {balance.currency}
        </Text>
        <FlatList
          data={transactions}
          style={styles.transactions}
          renderItem={({ item, index }) => (
            <TransactionRow
              id={item.id}
              amount={item.amount}
              counterparty={item.counterparty}
              category={item.category}
              currency={item.currency}
              isLastRow={index + 1 === transactions.length}
            />
          )}
        />
      </View>
    </ScrollView>
  ) : (
    <View style={[styles.container, styles.balanceContainer]}>
      <Text style={styles.balanceText}>Loading...</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  contentContainer: {
    paddingTop: 30,
  },
  balanceContainer: {
    alignItems: 'center',
    marginHorizontal: 50,
  },
  balanceText: {
    fontSize: 17,
    color: 'rgba(96,100,109, 1)',
    lineHeight: 24,
    textAlign: 'center',
    marginBottom: 20,
  },
  transactions: {
    width: '100%',
  },
});
