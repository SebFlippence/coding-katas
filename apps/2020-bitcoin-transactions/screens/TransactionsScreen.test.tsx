import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import TransactionsScreen from './TransactionsScreen';

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn().mockImplementation(() => ({
    addListener: jest.fn(),
  })),
}));

describe('TransactionsScreen', () => {
  const renderer = new ShallowRenderer();

  test('rendering a loading screen', () => {
    const tree = renderer.render(<TransactionsScreen />);
    expect(tree).toMatchSnapshot();
  });
});
