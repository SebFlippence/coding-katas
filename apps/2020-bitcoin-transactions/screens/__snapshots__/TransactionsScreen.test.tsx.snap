// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsScreen rendering a loading screen 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      },
      Object {
        "alignItems": "center",
        "marginHorizontal": 50,
      },
    ]
  }
>
  <Text
    style={
      Object {
        "color": "rgba(96,100,109, 1)",
        "fontSize": 17,
        "lineHeight": 24,
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Loading...
  </Text>
</View>
`;
